contracts/profiles/DNSResolver.sol

```diff
@@ -0,0 +1,142 @@
+        uint256 offset = 0;
+        bytes memory name;
+        bytes memory value;
+        bytes32 nameHash;
+        // Iterate over the data to add the resource records
+        for (RRUtils.RRIterator memory iter = data.iterateRRs(0); !iter.done(); iter.next()) {
+            if (resource == 0) {
+                resource = iter.dnstype;
+                name = iter.name();
+                nameHash = keccak256(abi.encodePacked(name));
+                value = bytes(iter.rdata());
+            } else {
+                bytes memory newName = iter.name();
+                if (resource != iter.dnstype || !name.equals(newName)) {
+                    setDNSRRSet(node, name, resource, data, offset, iter.offset - offset, value);
+                    resource = iter.dnstype;
+                    offset = iter.offset;
+                    name = newName;
+                    nameHash = keccak256(name);
+                    value = bytes(iter.rdata());
+                }
+            }
+        }
+        if (name.length > 0) {
+            setDNSRRSet(node, name, resource, data, offset, data.length - offset, value);
+        }
+    }
+
+    /**
+     * Obtain a DNS record.
+     * @param node the namehash of the node for which to fetch the record
+     * @param name the keccak-256 hash of the fully-qualified name for which to fetch the record
+     * @param resource the ID of the resource as per https://en.wikipedia.org/wiki/List_of_DNS_record_types
+     * @return the DNS record in wire format if present, otherwise empty
+     */
+    function dnsRecord(bytes32 node, bytes32 name, uint16 resource) public view returns (bytes memory) {
+        return records[node][versions[node]][name][resource];
+    }
+
+    /**
+     * Check if a given node has records.
+     * @param node the namehash of the node for which to check the records
+     * @param name the namehash of the node for which to check the records
+     */
+    function hasDNSRecords(bytes32 node, bytes32 name) public view returns (bool) {
+        return (nameEntriesCount[node][versions[node]][name] != 0);
+    }
+
+    /**
+     * Clear all information for a DNS zone.
+     * @param node the namehash of the node for which to clear the zone
+     */
+    function clearDNSZone(bytes32 node) public authorised(node) {
+        versions[node]++;
+        emit DNSZoneCleared(node);
+    }
+
+    function supportsInterface(bytes4 interfaceID) public pure returns(bool) {
+        return interfaceID == DNS_RECORD_INTERFACE_ID || super.supportsInterface(interfaceID);
+    }
+
+    function setDNSRRSet(
+        bytes32 node,
+        bytes memory name,
+        uint16 resource,
+        bytes memory data,
+        uint256 offset,
+        uint256 size,
+        bytes memory value) private
+    {
+        uint16 version = versions[node];
+        bytes32 nameHash = keccak256(name);
+        bytes memory rrData = data.substring(offset, size);
+        if (value.length == 0) {
```

Comment:
This seems to be the only use of `value`. Shouldn't this simply rely on `size` instead?

Reply:
`size` is the size of the record, which includes name, resource type _etc_.  I could change it to do the maths on the existing values but then we're adding complexity to remove a parameter, and I'd rather err on the side of simplicity.

Reply:
Couldn't you replace the parameter with, say, `bool delete`, then, since it's only used to check if the length is nonzero?

Reply:
Done.